src/btree/bt_page.c

```diff
@@ -693,57 +672,129 @@ __page_merge_deltas_with_base_image_old(WT_SESSION_IMPL *session, WT_REF *ref, W
-    key.data = unpack_key->data;
-    key.size = unpack_key->size;
-    key_prefix = unpack_key->prefix;
-    if (key_prefix == 0) {
-        lastkey->data = key.data;
-        lastkey->size = key.size;
-    } else {
-        WT_ASSERT(session, lastkey->size >= key_prefix);
-        lastkey->size = key_prefix;
-        WT_RET(__wt_buf_grow(session, lastkey, key_prefix + key.size));
-        memcpy((uint8_t *)lastkey->mem + key_prefix, key.data, key.size);
-        lastkey->size = key_prefix + key.size;
-    }
+    /* Decompress prefix compressed key. */
+    WT_RET(__wt_cell_decompress_prefix_key(
+      session, s->current_key, s->unpack_key->data, s->unpack_key->size, s->unpack_key->prefix));
 
     /*
      * Unpack the value if we have entries left, we may see a key unpacked if the entry has empty
      * value, in such case we set key_unpacked, and in the next run we can swap the pointers of
      * base_unpack_key and base_unpack_value so base_unpack_key can point to the right memory
      * location.
      */
-    if (i < entries) {
-        __wt_cell_unpack_kv(session, dsk, (WT_CELL *)cell, unpack_value);
-        cell += unpack_value->__len;
-        i++;
+    if (s->idx < s->entries) {
+        __wt_cell_unpack_kv(session, dsk, (WT_CELL *)s->cell, s->unpack_value);
+        s->cell += s->unpack_value->__len;
+        s->idx++;
         WT_ASSERT(session,
-          unpack_value->type != WT_CELL_KEY_OVFL && unpack_value->type != WT_CELL_VALUE_OVFL);
-        key_unpacked = unpack_value->type == WT_CELL_KEY;
+          s->unpack_value->type != WT_CELL_KEY_OVFL && s->unpack_value->type != WT_CELL_VALUE_OVFL);
+        s->key_unpacked = s->unpack_value->type == WT_CELL_KEY;
+    }
+
+    return (0);
+}
+
+/*
+ * __page_init_base_leaf_merge_state --
+ *     Initialize base leaf merge state.
+ */
+static int
+__page_init_base_leaf_merge_state(
+  WT_SESSION_IMPL *session, WT_BTREE *btree, WT_PAGE_HEADER *base_dsk, WTI_BASE_LEAF_MERGE_STATE *s)
+{
+    s->entries = base_dsk->u.entries;
+    s->idx = 0;
+    s->cell = WT_PAGE_HEADER_BYTE(btree, base_dsk);
+    s->found_next = false;
+    s->key_unpacked = false;
+
+    WT_RET(__wt_calloc_one(session, &s->unpack_key));
+    WT_RET(__wt_calloc_one(session, &s->unpack_value));
+    WT_RET(__wt_scr_alloc(session, 0, &s->current_key));
+
+    return (0);
+}
+
+/*
+ * __page_free_base_leaf_merge_state --
+ *     Free base leaf merge state.
+ */
+static void
+__page_free_base_leaf_merge_state(WT_SESSION_IMPL *session, WTI_BASE_LEAF_MERGE_STATE *s)
+{
+    __wt_free(session, s->unpack_key);
+    __wt_free(session, s->unpack_value);
+    __wt_scr_free(session, &s->current_key);
+}
+
+/*
+ * __page_init_delta_leaf_merge_state --
+ *     Initialize delta leaf merge state.
+ */
+static int
+__page_init_delta_leaf_merge_state(WT_SESSION_IMPL *session, WT_BTREE *btree, WT_ITEM *deltas,
+  size_t delta_size, WTI_DELTA_LEAF_MERGE_STATE ***sp)
+{
+    WTI_DELTA_LEAF_MERGE_STATE **s = NULL;
```

Comment:
Can this be WTI_DELTA_LEAF_MERGE_STATE *s so we don't need to allocate each entry?